authentication/oidc.go

```diff
@@ -367,3 +372,97 @@ func (a oidcAuthenticator) Middleware() Middleware {
 		})
 	}
 }
+
+//nolint:gocognit
+func (a oidcAuthenticator) GRPCMiddleware() grpc.StreamServerInterceptor {
+	return grpc_middleware_auth.StreamServerInterceptor(func(ctx context.Context) (context.Context, error) {
+		var token string
+
+		md, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			return ctx, status.Error(codes.Internal, "metadata error")
+		}
+
+		authorizationHeader := ""
+		authorizationHeaders := md["authorization"]
+		if len(authorizationHeaders) > 0 {
+			authorizationHeader = authorizationHeaders[0]
+		}
+
+		if authorizationHeader != "" {
+			authorization := strings.Split(authorizationHeader, " ")
+			if len(authorization) != 2 {
+				const msg = "invalid Authorization header"
+				level.Debug(a.logger).Log("msg", msg)
+				return ctx, status.Error(codes.InvalidArgument, msg)
+			}
+
+			token = authorization[1]
+		}
+
+		idToken, err := a.verifier.Verify(oidc.ClientContext(ctx, a.client), token)
```

Comment:
It looks like from this line below, the code is very similar with how HTTP middleware handles this. Do you think we could extract those parts and reuse a single method for both HTTP and gRPC middleware?